### PR TITLE
SONARJAVA-1265 TypeJavaSymbol.getFullyQualifiedName should return full qualified name for nested and anonymous classes

### PR DESCRIPTION
**Pull request by mpaladin**
https://jira.sonarsource.com/browse/SONARJAVA-1265

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java [hunk 1]
```diff
@@ -291,11 +292,37 @@ Scope completedMembers() {
...
+        internalName = name;
+      }
+    }
+
+    private String registerClassInternalName(String name) {
+      int num;
+      if (internalNames.containsKey(name)) {
+        num = internalNames.get(name) + 1;
+      } else {
+        num = 1;
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java]: initialize num to 1 and get rid of else.

File: java-squid/src/test/java/org/sonar/java/resolve/JavaTypeTest.java [hunk 2]
```diff
@@ -221,6 +228,46 @@ public void fully_qualified_name() throws Exception {
...
+
+  @Test
+  public void test_fully_qualified_name() {
+    File bytecodeDir = new File("target/test-classes");
+    ClassVisitor visitor = new ClassVisitor(bytecodeDir);
+    JavaAstScanner.scanSingleFile(
+      new File("src/test/java/org/sonar/java/resolve/targets/FullyQualifiedName.java"), new VisitorsBridge(visitor, Lists.newArrayList(bytecodeDir)));
+  }
+
+  private static class ClassVisitor extends SubscriptionVisitor {
```
Comment (benzonico) [inline java-squid/src/test/java/org/sonar/java/resolve/JavaTypeTest.java]: This does way more than just visiting class. ;) Please rename it.

File: java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java [hunk 3]
```diff
@@ -324,11 +351,17 @@ public Scope typeParameters() {
     }
 
     public String getFullyQualifiedName() {
-      String ownerName = "";
-      if (!owner.name.isEmpty()) {
-        ownerName = owner.name + ".";
+      String newQualification;
+      if (owner.isPackageSymbol()) {
+        newQualification = owner.name.isEmpty() ? "" : owner.name + ".";
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java]: MAJOR Add parentheses to make the operator precedence explicit. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS864)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java [hunk 4]
```diff
@@ -291,11 +292,37 @@ Scope completedMembers() {
     Scope typeParameters;
     List<JavaType.TypeVariableJavaType> typeVariableTypes;
     ClassTree declaration;
+    private final String internalName;
+    private final Map<String, Integer> internalNames = new HashMap<>();
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java]: Probably better to use MultiSet from guava.

File: java-squid/src/test/java/org/sonar/java/resolve/JavaSymbolTest.java [hunk 5]
```diff
@@ -80,21 +80,21 @@ public void test_TypeSymbol() {
 
   @Test
   public void access_to_superclass_should_trigger_completion() {
-    JavaSymbol.TypeJavaSymbol typeSymbol = spy(new JavaSymbol.TypeJavaSymbol(42, "name", null));
+    JavaSymbol.TypeJavaSymbol typeSymbol = spy(new JavaSymbol.TypeJavaSymbol(42, "name", new JavaSymbol.PackageJavaSymbol(null, null)));
```
Comment (benzonico) [inline java-squid/src/test/java/org/sonar/java/resolve/JavaSymbolTest.java]: Maybe this package symbol can be factorized.

**Pull request merged**